119 turinio prenumerata

adp-atea opened the pull request.

adp-atea committed:
add subscription template tag

sirex reviewed (changes requested):
Migrations does not work.

```
$ poetry run python manage.py migrate
  Applying vitrina_messages.0002_subscription...
Traceback (most recent call last):
[...]
  File "vitrina/messages/migrations/0002_subscription.py", line 14, in migrate_subscriptions
    content_type = ContentType.objects.get(app_label='vitrina_datasets', model='dataset')
  File "django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "django/db/models/query.py", line 435, in get
    raise self.model.DoesNotExist(
__fake__.ContentType.DoesNotExist: ContentType matching query does not exist.
```

adp-atea committed:
Merge branch 'devel' into 119-turinio-prenumerata

adp-atea committed:
fix dataset migration

adp-atea commented:
I fixed dataset initial migration and went through the whole process of creating docker image, migrating MySQL database to PostgeSQL, creating some test data for migration that gives errors and then running the migrate command. Everything went smoothly. Is it possible that in this case you didn't run contenttypes migrations or the database got corrupted somehow?

sirex committed:
Fix migration error

You can't user ContentType.objects.get(), because content type might not
be cached yet, you must use Conetenttype.objects.get_for_model().

See: https://docs.djangoproject.com/en/3.2/ref/contrib/contenttypes/#the-contenttypemanager

sirex committed:
Add Subscription to admin

sirex committed:
Fix access to project details page

I was unable to access project details page, because:

- there was no menu entry to project list,
- project in project list did not have link to details page,
- details page title was not shown.

sirex commented on vitrina/messages/models.py:
```diff
@@ -82,3 +86,17 @@ class UserSubscription(models.Model):
+    user = models.ForeignKey(User, models.CASCADE)
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey('content_type', 'object_id')
```
`user`, `content_type` and `object_id` should be [unique_together](https://docs.djangoproject.com/en/3.2/ref/models/options/#unique-together), to prevent duplicate subscriptions.

sirex reviewed (changes requested):
> I fixed dataset initial migration and went through the whole process of creating docker image, migrating MySQL database to PostgeSQL, creating some test data for migration that gives errors and then running the migrate command. Everything went smoothly. Is it possible that in this case you didn't run contenttypes migrations or the database got corrupted somehow?

Fixed this in https://github.com/atviriduomenys/katalogas/pull/164/commits/2f7dc5c346dc85eaec3cbafd073aeaaae586edd7, the issue was, that you tried to access `ContentType` directly, but `get_for_model()` manager method should be used, because `ContentType` uses database as a cache and records might or might not be cached, especially during migrations.

adp-atea committed:
added unique together
